arangod/Aql/DocumentProducingHelper.cpp

```diff
@@ -454,6 +475,81 @@ IndexIterator::CoveringCallback aql::getCallback(
   };
 }
 
+template<bool checkUniqueness, bool skip>
+IndexIterator::CoveringCallback aql::getCallback(
+    DocumentProducingCallbackVariant::WithFilterCoveredByIndex,
+    DocumentProducingFunctionContext& context) {
+  return [&context](LocalDocumentId const& token,
+                    IndexIteratorCoveringData& covering) {
+    // this must only be used if we have a filter!
+    TRI_ASSERT(context.hasFilter());
+    TRI_ASSERT(context.getAllowCoveringIndexOptimization());
+    if constexpr (checkUniqueness) {
+      if (!context.checkUniqueness(token)) {
+        // Document already found, skip it
+        return false;
+      }
+    }
+
+    context.incrScanned();
+
+    // recycle our Builder object
+    VPackBuilder& objectBuilder = context.getBuilder();
+    objectBuilder.clear();
+    objectBuilder.openObject(true);
+
+    // projections from the index for the filter condition
+    context.getFilterProjections().toVelocyPackFromIndex(
+        objectBuilder, covering, context.getTrxPtr());
+
+    objectBuilder.close();
+
+    if (!context.checkFilter(objectBuilder.slice())) {
```

Comment:
We have to copy the values from the index into a temp object for now, as we still have to run arbitrary AQL expressions on it. And our expression evaluation functionality works by accessing attributes from objects by name, e.g. `FILTER ABS(doc.value3.sub) != 23`: here it will need to extract the `sub` attribute from a value3 `attribute` from an object. And the object thus needs to contain the required data.
I agree this is not ideal, but we would need to rewrite the expression evaluation so it can fetch data from somewhere else. Probably not a bad idea, but IMO out of scope for this PR and 3.10.

Reply:
> out of scope for this PR and 3.10
Agreed! This was more an informative question, I figured this would not be changed in this PR.